File: frontend/src/services/competitions.ts
import { api } from '@/services/api';
import type { MemeContainerResponse } from '@/services/instagram';
import type { MemeResponse } from '@/services/memes';

export type CompetitionPeriodType = 'day' | 'week' | 'month';

// Native memes and MemeContainers (Instagram Companion Mode) compete together in one
// ranking — see backend services/competitions.py::_standings_query. Standings are ranked
// by net vote score (upvotes minus downvotes cast within the period window), derived
// automatically from the same upvote/downvote votes cast via /memes/{id}/votes and
// /instagram/containers/{id}/votes — there's no separate "cast a competition vote" action.
// `meme` is null only for a *closed period's* winner whose post was deleted after already
// winning — the winner slot/score stay fixed (deletion never rewrites who actually won),
// but there's no live content left to show (the Cloudinary asset is gone). `is_deleted`
// tells the UI to render "Deleted Post" and disable the click-through. A meme entry in the
// *live* current-standings list is never in this state — a deleted post is excluded from
// that ranking outright (see backend services/competitions.py).
export type StandingContent =
  | { kind: 'meme'; meme: MemeResponse | null; is_deleted: boolean }
  | { kind: 'container'; container: MemeContainerResponse };

export interface StandingEntryResponse {
  rank: number;
  content: StandingContent;
  score: number;
}

export interface StandingsPageResponse {
  period_type: CompetitionPeriodType;
  period_key: string;
  is_closed: boolean;
  items: StandingEntryResponse[];
}

export interface WinnerResponse {
  period_type: CompetitionPeriodType;
  period_key: string;
  content: StandingContent | null;
  score: number;
}

export function getCurrentStandingsRequest(periodType: CompetitionPeriodType, limit = 20) {
  return api.get<StandingsPageResponse>(`/competitions/${periodType}/current`, { limit });
}

export function getWinnerRequest(periodType: CompetitionPeriodType, periodKey: string) {
  return api.get<WinnerResponse>(`/competitions/${periodType}/winner`, { period_key: periodKey });
}

File: backend/app/schemas/competitions.py
from typing import Literal

from pydantic import BaseModel

from app.models.competition_period import CompetitionPeriod
from app.schemas.instagram import MemeContainerOut
from app.schemas.memes import MemeOut


class StandingContentMeme(BaseModel):
    kind: Literal["meme"]
    # Null when the meme was deleted after already deciding a *closed* period's winner —
    # its score/rank still stand (a deletion never rewrites who actually won), but its
    # content is gone (the Cloudinary asset is cleaned up on delete, so there's nothing
    # live to show) and `is_deleted` tells the client to render "Deleted Post" and disable
    # the click-through instead of a broken image. Never null for a *live* period's
    # standings — a deleted meme is excluded from those entirely (see services/competitions.py).
    meme: MemeOut | None
    is_deleted: bool = False


class StandingContentContainer(BaseModel):
    kind: Literal["container"]
    container: MemeContainerOut


class StandingEntry(BaseModel):
    rank: int
    content: StandingContentMeme | StandingContentContainer
    score: int


class StandingsPage(BaseModel):
    period_type: CompetitionPeriod
    period_key: str
    is_closed: bool
    items: list[StandingEntry]


class WinnerOut(BaseModel):
    period_type: CompetitionPeriod
    period_key: str
    content: StandingContentMeme | StandingContentContainer | None
    score: int
